Rmlui integration fork floofy

ChrisFloofyKitsune opened the pull request:
just shoving my progress in here every now and then

cleaned up the code that keeps track of the Rml UI contexts so that it's not bullying RmlSolLua so much

ChrisFloofyKitsune committed:
add couple things to .gitignore
make lua binding for RemoveContext

ChrisFloofyKitsune committed:
add couple things to .gitignore
make lua binding for RemoveContext

ChrisFloofyKitsune committed:
do less bullying of the RmlSolLua lib
instead set up our own plugin to keep track of created/removed contexts

ChrisFloofyKitsune committed:
do less bullying of the RmlSolLua lib
instead set up our own plugin to keep track of created/removed contexts

ChrisFloofyKitsune committed:
Merge remote-tracking branch 'chrisfloofykitsune/rmlui-integration-fork-floofy' into rmlui-integration-fork-floofy

loveridge commented:
Took a look and the changes look good! I'll merge these if you mark it ready.

ChrisFloofyKitsune committed:
attempt to add thread safety to RmlGui::data->context accesses
allow RmlUI to properly set cursors
have RmlUI input events take precedence after other inputs (but after regular Lua input)

p2004a commented:
This PR mixes some unrelated to rmlui code refactoring in engine with rmlui related changes. You should not do that, it makes it very painful to review.

Any unrelated refactoring should be split out.

p2004a commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -2740,6 +2739,10 @@ int LuaUnsyncedCtrl::WarpMouse(lua_State* L)
 int LuaUnsyncedCtrl::SetMouseCursor(lua_State* L)
 {
+	// ignore lua requests to control cursor when mouse is over RmlUI element
+	if (RmlGui::IsActive())
```
If it's already possible in Lua that one widget changes cursor when mouse is over a different one, then this change is a breaking behavior.

ChrisFloofyKitsune replied:
The widget that's caused me trouble- that made me include that line is `cmd_rclick_quick_build_resource_extractor.lua` because it sets the cursor when you're hovering over a metal spot

Looking into things more- it seems that the last widget to set the cursor "wins".

All the gui widgets seem to have logic along the lines of  "if the mouse is over me, set the cursor to 'cursornormal' every frame" in either `widget:Update()` or `widget:DrawScreen()`

RmlUI only calls SetMouseCursor when it determines internally that there's been change in shown cursor caused by hovering over something. That results in the cursor changing for one frame before a widget sets it to something else.

It does **not** act like the widgets which fight over the cursor icon every frame.
That check makes the widgets fighting over the mouse cursor get completely ignored.

Either I do this or I refactor a bunch of widgets. 😑

ChrisFloofyKitsune replied:
Also, now that I think about, `IsActive()` is a really bad name now for what that check does.

I'll rename that to `IsMouseAbove()` or some such

ChrisFloofyKitsune replied:
I could also do a thing where I track the last requested cursor from RmlUi

This works because RmlUI sets the cursor to an empty string as the last-case scenario

https://mikke89.github.io/RmlUiDoc/pages/cpp_manual/contexts.html#mouse-cursor

that way an Rml element could just not do anything to the cursor icon- letting Widgets and other things have control

ChrisFloofyKitsune replied:
I've used that implementation detail to refactor how the mouse cursor eventually gets set by RmlUi
If Rml is being interacted with and it wants a certain cursor- the CMouseHandler::Update() code will respect that.

ChrisFloofyKitsune commented:
> This PR mixes some unrelated to rmlui code refactoring in engine with rmlui related changes. You should not do that, it makes it very painful to review.
> 
> Any unrelated refactoring should be split out.

Sorry, I will be more careful of this in the future.
The only file I let my IDE format completely is RmlUi_Backend.cpp which... yeah- does not look great in GitHub.

ChrisFloofyKitsune committed:
have RmlGui store what cursor it would like (or "" to cede control of it) and have the system check that.
rename own custom Platform_SDL stuff to Platform_RTS
clean up code here and there, undo changes to random files

ChrisFloofyKitsune commented on rts/Game/UI/InputReceiver.cpp:
```diff
@@ -68,10 +69,14 @@ void CInputReceiver::DrawReceivers()
 CInputReceiver* CInputReceiver::GetReceiverAt(int x, int y)
 {
-	// always ask Lua first
+	// check lua first
 	if (luaInputReceiver != nullptr && luaInputReceiver->IsAbove(x, y))
 		return luaInputReceiver;
 
+	// check RmlUI second
+	if (RmlGui::IsMouseInteractingWith())
+		return RmlGui::GetInputReceiver();
+
```
I think we might want to be checking RmlGui first?

loveridge replied:
Perhaps. In other places rmlui gets first shot at handling events, so maybe it should here too.

loveridge commented on rts/Rml/Backends/RmlUi_Backend.h:
```diff
@@ -29,15 +29,12 @@
-#include <RmlUi/Core/Input.h>
-#include <RmlUi/Core/RenderInterface.h>
-#include <RmlUi/Core/SystemInterface.h>
+#include <RmlUI/Core.h>
```
`RmlUi`

ChrisFloofyKitsune replied:
asdfdsagfdhf

ChrisFloofyKitsune replied:
will fix

ChrisFloofyKitsune committed:
Platform_RTS ended up being disliked, so re-renamed stuff to "RmlSystemInterface" and the namespace to "RmlSDLSpring"

ChrisFloofyKitsune committed:
undo refactoring of import lists in non-Rml files
improve a comment

ChrisFloofyKitsune committed:
check Rml input first

ChrisFloofyKitsune committed:
fix a singular character to be lowercase blargh
